### PR TITLE
[CM-1592] Support for logging of assets folder

### PR DESCRIPTION
**Pull request by yaricom**
### comet-java-client
- Added support for the form parameters/fields when uploading files
- Refined Experiment and OnlineExperiment interfaces to properly handle context
- Added support of form fields/params for multipart requests
- Added ApiExperiment interface to make this type of supported experiments
- Introduced BaseExperimentAsync as superclass for asynchronous experiment implementations
- Introduced resource bundle to maintain user facing strings
- Introduced ExperimentContext to maintain context of the experiment: step, epoch, context ID
- Fixed logCode and uploadAsset to use asynchronous upload operations in OnlineExperiment
- Implemented logAssetFolder allowing to upload all asset files from specific directory

### comet-examples
- Updated OnlineExperimentExample to include example of assets folder logging
- Updated OnlineExperimentExample to demonstrate usage of the ExperimentContext
- Updated OnlineExperimentExample to demonstrate logCode usage

### CODE REVIEW

File: comet-java-client/src/main/java/ml/comet/experiment/impl/OnlineExperimentBuilderImpl.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        if (StringUtils.isEmpty(this.baseUrl)) {
+            this.baseUrl = COMET_BASE_URL.getString();
+        }
+        if (this.maxAuthRetries == -1) {
+            this.maxAuthRetries = COMET_MAX_AUTH_RETRIES.getInt();
+        }
+        Duration cleaningTimeout = COMET_TIMEOUT_CLEANING_SECONDS.getDuration();
+
+        return new OnlineExperimentImpl(
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/OnlineExperimentBuilderImpl.java]: Fixed

File: comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Describes context of the {@link ml.comet.experiment.Experiment}.
+ */
+@Data
+public final class ExperimentContext {
+    private long step;
+    private long epoch;
+    private String context;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java:13]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+import lombok.Data;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Describes context of the {@link ml.comet.experiment.Experiment}.
+ */
+@Data
+public final class ExperimentContext {
+    private long step;
+    private long epoch;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java:12]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+import lombok.Data;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Describes context of the {@link ml.comet.experiment.Experiment}.
+ */
+@Data
+public final class ExperimentContext {
+    private long step;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java:11]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ * Describes asset data.
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Asset {
+    private File file;
+    private byte[] fileLikeData;
+    private String fileName;
+    private AssetType type;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:21]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Map;
+
+/**
+ * Describes asset data.
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Asset {
+    private File file;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:18]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    private byte[] fileLikeData;
+    private String fileName;
+    private AssetType type;
+    private boolean overwrite;
+    private long step;
+    private long epoch;
+    private String groupingName;
+    private Map<String, Object> metadata;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:26]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    private String assetId;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:27]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    private String fileExtension;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:28]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Describes asset data.
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Asset {
+    private File file;
+    private byte[] fileLikeData;
+    private String fileName;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:20]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    private AssetType type;
+    private boolean overwrite;
+    private long step;
+    private long epoch;
+    private String groupingName;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:25]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Asset {
+    private File file;
+    private byte[] fileLikeData;
+    private String fileName;
+    private AssetType type;
+    private boolean overwrite;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:22]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    private long step;
+    private long epoch;
+    private String groupingName;
+    private Map<String, Object> metadata;
+    private String assetId;
+    private String fileExtension;
+    private String context;
+    private boolean remote;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:30]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * Describes asset data.
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Asset {
+    private File file;
+    private byte[] fileLikeData;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:19]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    private String fileName;
+    private AssetType type;
+    private boolean overwrite;
+    private long step;
+    private long epoch;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:24]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Asset {
+    private File file;
+    private byte[] fileLikeData;
+    private String fileName;
+    private AssetType type;
+    private boolean overwrite;
+    private long step;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:23]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/LogDataResponse.java [hunk 17]
```diff
@@ -18,6 +18,7 @@
     private String msg;
     private int code;
     private int sdkErrorCode;
+    private String data;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/LogDataResponse.java:21]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    public ExperimentContext(long step, long epoch) {
+        this(step, epoch, StringUtils.EMPTY);
+    }
+
+    /**
+     * Creates new instance with specified parameters.
+     *
+     * @param step the current step of the experiment.
+     */
+    public ExperimentContext(long step) {
```
Comment (Nimrod007) [inline comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java:46]: why not use step and epoch as Long (non primitive) so we dont need to default to 0 and we can default to null ?

File: comet-java-client/src/main/java/ml/comet/experiment/impl/ApiExperimentImpl.java [hunk 19]
```diff
@@ -131,7 +132,7 @@ private ApiExperimentBuilderImpl() {
         }
 
         @Override
-        public ApiExperimentImpl build() {
+        public ApiExperiment build() {
```
Comment (Nimrod007) [inline comet-java-client/src/main/java/ml/comet/experiment/impl/ApiExperimentImpl.java:135]: nice!

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+import java.io.File;
+import java.util.Map;
+
+/**
+ * Describes asset data.
+ */
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class Asset {
```
Comment (Nimrod007) [inline comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:17]: for non mandatory fields (step, epoch) I think its cleaner to use null as default (and use non primitives)

File: comet-java-client/src/test/java/ml/comet/experiment/impl/OnlineExperimentTest.java [hunk 21]
```diff
@@ -58,6 +65,63 @@
...
+        //
+        ExperimentContext context = new ExperimentContext(123, 1042, "train");
+        OnCompleteAction onComplete = new OnCompleteAction();
+        experiment.logAssetFolder(root.toFile(), false, true, context, onComplete);
+
+        awaitForCondition(onComplete, "log assets' folder timeout", 60);
+
+        // wait for assets become available and validate results
+        //
+        awaitForCondition(() -> experiment.getAssetList(ASSET_TYPE_ALL).size() == 5, "Assets was uploaded");
```
Comment (Nimrod007) [inline comet-java-client/src/test/java/ml/comet/experiment/impl/OnlineExperimentTest.java:112]: test is valid. its better to get the expected (in this case `5` ) dynamic and not hard code this. for the reader i dont have all the information to understand the test by reading it and will require to go to the root.toFile() and see that we actually have 5 files in it. better to scan list and get size of list on assertion. this way the reader can understand everything in the test

File: comet-java-client/src/test/java/ml/comet/experiment/impl/OnlineExperimentTest.java [hunk 22]
```diff
@@ -323,25 +388,48 @@ public void testLogAndGetExperimentTime() {
 
     @Test
     public void testUploadAndGetAssets() {
-        OnlineExperiment experiment = createOnlineExperiment();
+        OnlineExperimentImpl experiment = (OnlineExperimentImpl) createOnlineExperiment();
 
+        // Make sure experiment has no assets
```
Comment (Nimrod007) [inline comet-java-client/src/test/java/ml/comet/experiment/impl/OnlineExperimentTest.java:393]: when the action is very strait forward like x.isEmpty no need to document (its self explained) 

comments / documents are good on complex logic on non trivial assertion / setup

File: comet-java-client/src/test/java/ml/comet/experiment/impl/utils/AssetUtilsTest.java [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+        assertEquals(asset.getFileName(), subFolderFile.getFileName().toString(), "wrong asset file name");
+        assertEquals("txt", asset.getFileExtension(), "wrong file extension");
+    }
+
+    @Test
+    public void testWalkFolderAssets() throws IOException {
+        // tests that correct number of assets returned
+        Stream<Asset> assets = AssetUtils.walkFolderAssets(
+                root.toFile(), true, true);
+        assertEquals(5, assets.count(), "wrong assets count");
```
Comment (Nimrod007) [inline comet-java-client/src/test/java/ml/comet/experiment/impl/utils/AssetUtilsTest.java:59]: try to fetch the actual 5 from a dynamic list (if some1 adds a file to resource folder all tests will break)

**Review by Nimrod007 — CHANGES_REQUESTED** [review]

File: comet-java-client/src/test/java/ml/comet/experiment/impl/utils/AssetUtilsTest.java [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class AssetUtilsTest {
+    private static Path root;
+    private static Path subFolderFile;
+
+    @BeforeAll
+    static void setup() throws IOException {
+        // create temporary directory tree
+        root = Files.createTempDirectory("testFileUtils");
```
Comment (Nimrod007) [inline comet-java-client/src/test/java/ml/comet/experiment/impl/utils/AssetUtilsTest.java:28]:
```suggestion
        root = Files.createTempDirectory("testAssetUtils");
```

File: comet-java-client/src/test/java/ml/comet/experiment/impl/OnlineExperimentTest.java (hunk 21 above)
Comment (yaricom) [reply comet-java-client/src/test/java/ml/comet/experiment/impl/OnlineExperimentTest.java]: Fixed

File: comet-java-client/src/test/java/ml/comet/experiment/impl/utils/AssetUtilsTest.java (hunk 23 above)
Comment (yaricom) [reply comet-java-client/src/test/java/ml/comet/experiment/impl/utils/AssetUtilsTest.java]: Fixed

File: comet-java-client/src/test/java/ml/comet/experiment/impl/utils/AssetUtilsTest.java (hunk 24 above)
Comment (yaricom) [reply comet-java-client/src/test/java/ml/comet/experiment/impl/utils/AssetUtilsTest.java]: Fixed.

File: comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java (hunk 18 above)
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java:46]: Fixed

File: comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+import lombok.Data;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Describes context of the {@link ml.comet.experiment.Experiment}.
+ */
+@Data
+public final class ExperimentContext {
+    private Long step;
+    private Long epoch;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java:12]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+    private File file;
+    private byte[] fileLikeData;
+    private String fileName;
+    private AssetType type;
+    private Boolean overwrite;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:22]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java [hunk 27]
```diff
@@ -0,0 +1,120 @@
...
+
+import lombok.Data;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Describes context of the {@link ml.comet.experiment.Experiment}.
+ */
+@Data
+public final class ExperimentContext {
+    private Long step;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/context/ExperimentContext.java:11]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+    private Long step;
+    private Long epoch;
+    private String groupingName;
+    private Map<String, Object> metadata;
+    private String assetId;
+    private String fileExtension;
+    private String context;
+    private Boolean remote;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/asset/Asset.java:30]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/ExperimentAssetLink.java [hunk 29]
```diff
@@ -16,9 +16,9 @@
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class ExperimentAssetLink {
     private String fileName;
-    private long fileSize;
+    private Long fileSize;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/ExperimentAssetLink.java:19]: @sonatype-lift ignore

File: comet-java-client/src/main/java/ml/comet/experiment/model/ExperimentAssetLink.java [hunk 30]
```diff
@@ -16,9 +16,9 @@
...
     private String runContext;
-    private Integer step;
+    private Long step;
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/model/ExperimentAssetLink.java:21]: @sonatype-lift ignore

**Comment by yaricom** [comment]
The example experiment output
https://www.comet.ml/test-javasdk/java-sdk/d174bc0088e64e4b994956827599f200?experiment-tab=assets

**Review by Nimrod007 — APPROVED** [review]

**Review by Lothiraldan — APPROVED** [review]
LGTM, only one small typo

File: comet-java-client/src/main/java/ml/comet/experiment/impl/utils/AssetUtils.java [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.stream.Stream;
+
+/**
+ * Utilities to work with assets.
+ */
+@UtilityClass
+public class AssetUtils {
+
+    /**
+     * Walks through th asset files in the given folder and produce stream of {@link Asset} objects holding information
```
[the remark this answers is not in the document]
Comment (yaricom) [reply comet-java-client/src/main/java/ml/comet/experiment/impl/utils/AssetUtils.java]: Fixed

**Pull request merged**